### PR TITLE
Fix decorator metadata references to type-only-imported namespaces

### LINKED ISSUE #42624: emitDecoratorMetadata causes runtime errors by referencing type-only imports with namespaces
# Bug Report

Related to #37672 and #39337 - seems this bug still happens if class is wrapped by a namespace.

### 🔎 Search Terms

emitDecoratorMetadata, type-only imports, decorators

### 🕗 Version & Regression Information

Typescript 4.1.3, Typescript 4.2.0 beta

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about emitDecoratorMetadata and type-only imports

### ⏯ Playground Link

[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?esModuleInterop=false&target=99&strict=false#code/PTAEAEFMA8AdIE4EsC2kB2AXAhgGwCKQDGA9gtpmQM4BQIEkKSmhp5lCAspDgCYXY6YcFUzIimAFygAZniqQaS+uBlJcGbGmkKEANyRFIVAHSZaMWGUyh0W47GxHQAZUQGjVUAG8aoUJbWoES42FRebvqGkD4AvjTxQhBqGnbaoEjovDBmtKhWCDaYAJ7wPq7u0V6xsggkKKAA5CbAuh7GjQDcStkh2AgxpOiioL1kFGTS2OjF3YGFwaHhoJzYmT5+EGPsZAAUAJSbapC4vNKR7aYX0d2xQA)

### 💻 Code

```ts repro
// @experimentalDecorators
// @emitDecoratorMetadata
// @strict: false


// @filename: services.ts
export namespace Services {
  export class Service {}
}

// @filename: index.ts
import type { Services } from './services';

declare const decorator: any;
export class Main {
  @decorator()
  field: Services.Service;
}
```

### 🙁 Actual behavior

Typescript will reference "Services.Service" even though it's not imported (only the type was). This will crash in runtime:

```
__decorate([
decorator(),
__metadata("design:type", Services.Service)
], Main.prototype, "field", void 0);

```

### 🙂 Expected behavior

Should not reference non imported classes:

```
__decorate([
decorator(),
__metadata("design:type", Object)
], Main.prototype, "field", void 0);

```

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #42624

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -40348,9 +40348,14 @@ namespace ts {
...
             // Resolve the symbol as a value to ensure the type can be reached at runtime during emit.
+            let isTypeOnly = false;
+            if (isQualifiedName(typeName)) {
+                const rootValueSymbol = resolveEntityName(getFirstIdentifier(typeName), SymbolFlags.Value, /*ignoreErrors*/ true, /*dontResolveAlias*/ true, location);
+                isTypeOnly = rootValueSymbol?.declarations?.every(isTypeOnlyImportOrExportDeclaration) || false;
+            }
             const valueSymbol = resolveEntityName(typeName, SymbolFlags.Value, /*ignoreErrors*/ true, /*dontResolveAlias*/ true, location);
-            const isTypeOnly = valueSymbol?.declarations?.every(isTypeOnlyImportOrExportDeclaration) || false;
             const resolvedSymbol = valueSymbol && valueSymbol.flags & SymbolFlags.Alias ? resolveAlias(valueSymbol) : valueSymbol;
+            isTypeOnly ||= valueSymbol?.declarations?.every(isTypeOnlyImportOrExportDeclaration) || false;
```
Comment (sandersn) [inline src/compiler/checker.ts:40358]: Is the trailing `|| false` just needed to convert `undefined` to `false`? I think it's clearer to use `!!`
```suggestion
            isTypeOnly ||= !!valueSymbol?.declarations?.every(isTypeOnlyImportOrExportDeclaration);
```

(and above)

**Review by sandersn — APPROVED** [review]
Seems good. 1 minor style suggestion.

**Comment by stevekanter-arcticleaf** [comment]
@andrewbranch thanks so much for this! This has been an issue for me in a pretty large-scale application and has prevented me using `import type` to prevent errors. Any clue what version of TS this is going to be released in?

**Comment by andrewbranch** [comment]
4.4.2 (RC), along with nightlies starting tonight.

**Pull request merged**